crates/sui-core/src/authority/consensus_tx_status_cache.rs

```diff
@@ -0,0 +1,316 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use parking_lot::RwLock;
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::time::Duration;
+use sui_types::error::SuiError;
+use tokio::time::sleep;
+use tracing::debug;
+
+use mysten_common::sync::notify_read::NotifyRead;
+
+use crate::wait_for_effects_request::ConsensusTxPosition;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub(crate) enum ConsensusTxStatus {
+    // Transaction is voted to accept by a quorum of validators on fastpath.
+    FastpathCertified,
+    // Transaction is rejected, either by a quorum of validators or indirectly post-commit.
+    Rejected,
+    // Transaction is finalized post commit.
+    Finalized,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) enum NotifyReadConsensusTxStatusResult {
+    // The consensus position to be read has been updated with a new status.
+    Status(ConsensusTxStatus),
+    // The consensus position to be read has expired.
+    // Provided with the last committed round that was used to check for expiration.
+    Expired(u64),
+}
+
+#[derive(Default)]
+pub(crate) struct ConsensusTxStatusCache {
+    inner: RwLock<Inner>,
+    status_notify_read: NotifyRead<ConsensusTxPosition, ConsensusTxStatus>,
+    /// The depth of the garbage collection.
+    /// We use this to expire positions from old rounds.
+    gc_depth: u64,
+}
+
+#[derive(Default)]
+struct Inner {
+    /// A map of transaction position to its status from consensus.
+    transaction_status: HashMap<ConsensusTxPosition, ConsensusTxStatus>,
+    /// A map of consensus round to all transactions that were updated in that round.
+    round_lookup_map: BTreeMap<u64, HashSet<ConsensusTxPosition>>,
```

Comment:
It seems `round_lookup_map` is only used for GC. I think using BTreeMap for `transaction_status` would allow GC as well without this extra map, since ConsensusTxPosition is first ordered by round.